src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/SubjectAlternativeNameBuilder.cs

```diff
@@ -3,63 +3,109 @@
+            AddGeneralName(new GeneralNameAsn { DnsName = s_idnMapping.GetAscii(dnsName) });
         }
 
         public void AddUri(Uri uri)
         {
             if (uri == null)
                 throw new ArgumentNullException(nameof(uri));
 
-            _encodedTlvs.Add(_generalNameEncoder.EncodeUri(uri));
+            AddGeneralName(new GeneralNameAsn { Uri = uri.AbsoluteUri.ToString() });
         }
 
         public void AddIpAddress(IPAddress ipAddress)
         {
             if (ipAddress == null)
                 throw new ArgumentNullException(nameof(ipAddress));
 
-            _encodedTlvs.Add(_generalNameEncoder.EncodeIpAddress(ipAddress));
+            AddGeneralName(new GeneralNameAsn { IPAddress = ipAddress.GetAddressBytes() });
         }
 
         public void AddUserPrincipalName(string upn)
         {
             if (string.IsNullOrEmpty(upn))
                 throw new ArgumentOutOfRangeException(nameof(upn), SR.Arg_EmptyOrNullString);
 
-            _encodedTlvs.Add(_generalNameEncoder.EncodeUserPrincipalName(upn));
+            byte[] otherNameValue;
+            using (AsnWriter writer = new AsnWriter(AsnEncodingRules.DER))
+            {
+                Asn1Tag explicit0 = new Asn1Tag(TagClass.ContextSpecific, 0, isConstructed: true);
+                writer.PushSequence(explicit0);
+                writer.WriteCharacterString(UniversalTagNumber.UTF8String, upn);
+                writer.PopSequence(explicit0);
+                otherNameValue = writer.Encode();
+            }
+
+            OtherNameAsn otherName = new OtherNameAsn
+            {
+                TypeId = Oids.UserPrincipalName,
+                Value = otherNameValue,
+            };
+
+            AddGeneralName(new GeneralNameAsn { OtherName = otherName });
         }
 
         public X509Extension Build(bool critical=false)
         {
-            return new X509Extension(
-                Oids.SubjectAltName,
-                DerEncoder.ConstructSequence(_encodedTlvs),
-                critical);
+            using (AsnWriter writer = new AsnWriter(AsnEncodingRules.DER))
+            {
+                writer.PushSequence();
+                foreach (byte[] encodedName in _encodedNames)
+                {
+                    writer.WriteEncodedValue(encodedName);
+                }
+                writer.PopSequence();
+
+                return new X509Extension(
+                    Oids.SubjectAltName,
+                    writer.Encode(),
+                    critical);
+            }
+        }
+
+        private void AddGeneralName(GeneralNameAsn generalName)
+        {
+            try
+            {
+                // Verify that the general name can be serialized and store it.
+                using (AsnWriter writer = AsnSerializer.Serialize(generalName, AsnEncodingRules.DER))
+                {
+                    _encodedNames.Add(writer.Encode());
+                }
+            }
+            catch (EncoderFallbackException)
```

Comment:
Is EncoderFallbackException really what's coming out?  I would have expected the serializer to wrap it; but maybe that only happens during deserialize?

Reply:
I would have expected it too, but it doesn't happen at the moment. I was planning to open an issue for it.

Reply:
@bartonjs I was about to fix it in `AsnWriter` and submit PR when I realized that there are unit tests explicitly calling out this exception behavior - `VerifyWrite_Span_NonEncodable` and `VerifyWrite_String_NonEncodable`. Should I update the unit tests and submit the PR or drop it and consider this expected behavior?

Reply:
I had mixed feelings about your reasoning. On one hand it makes perfect sense to throw most appropriate exceptions in the AsnWriter and do the wrapping in the serializer. On the other hand `WriteInteger[Unsigned]` kinda goes against the rule by throwing `CryptographicException` instead of some kind of `ArgumentException`. Then again, I am really not a fan of `WriteInteger[Unsigned]` with the `ReadOnlySpan<byte>` argument. Nearly all the cases where it is used do explicit normalization (removing padding and/or adding zero to make the number unsigned) before calling the function. It feels so awkward to use that I believe it would make more sense if the normalization was part of `AsnWriter` (and the Unsigned option exposed by the serializer). It is easy to get it wrong and the API should be designed to avoid making mistakes.

That is, however, separate issue that deserves proper discussion. How should I proceed on this one? Should I fix the serializer to wrap the exception? If yes, should I do it as part of this PR, or a separate one?

Reply:
We can leave this as it is, though it'd be great if you filed an issue for the serializer exception, and if you want to have a discussion on the AsnWriter integers, a separate issue for that.

Reply:
Will file issues tomorrow. I wanted to finish removing all `DerEncoder` / `DerSequenceReader` usages first to get more acquainted with the APIs.